File: packages/ts/test/accounts/unified.test.ts
import { describe, it, expect } from "vitest";
import {
  deriveUnifiedAccount,
  unifiedAccountFromSeed,
  addressesFrom20,
  qoreAddresses,
} from "../../src/accounts/unified";
import { isValidBech32 } from "../../src/utils/address";

/**
 * Well-known PUBLIC test mnemonic (the canonical Hardhat/Anvil dev mnemonic).
 * NEVER use a real/funded mnemonic in tests.
 */
const TEST_MNEMONIC =
  "test test test test test test test test test test test junk";

/** Known-answer expectations for index 0 (asserted byte-for-byte). */
const KAT0 = {
  cosmos: "qor17w0adeg64ky0daxwd2ugyuneellmjgnxhkv37z",
  evm: "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266",
  svm: "HQ1S8pxTw4YN41GPdWYPQVfweQveXAUtfFnZNmvPfrYf",
  addr20: "f39fd6e51aad88f6f4ce6ab8827279cfffb92266",
};

/** KAT for the seed = 32×0x01 case. */
const KAT_SEED = {
  cosmos: "qor1rfjz7r3u8t65teavh5utquj3kwvsj983hh5zaj",
  evm: "0x1a642f0E3c3aF545E7AcBD38b07251B3990914F1",
  svm: "2n2Cnc7fib6rm4Azo1mbvMuHB3iCb1J254kEQDcrHyPu",
};

function toHex(bytes: Uint8Array): string {
  return Array.from(bytes)
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

describe("deriveUnifiedAccount", () => {
  it("matches the index-0 known-answer addresses exactly", async () => {
    const a = await deriveUnifiedAccount(TEST_MNEMONIC, 0);
    expect(a.cosmos).toBe(KAT0.cosmos);
    expect(a.evm).toBe(KAT0.evm);
    expect(a.svm).toBe(KAT0.svm);
    expect(toHex(a.addressBytes)).toBe(KAT0.addr20);
  });

  it("derives the deterministic ML-DSA-87 keypair (2592/4896) with a fixed prefix", async () => {
    const a = await deriveUnifiedAccount(TEST_MNEMONIC, 0);
    expect(a.pqc.publicKey.length).toBe(2592);
    expect(a.pqc.secretKey.length).toBe(4896);
    // Cross-language PQC determinism: the first 8 bytes are a known answer
    // (locked against the reference wallet-adapter).
    expect(toHex(a.pqc.publicKey.slice(0, 8))).toBe("4a685622f2a99d54");
    // deterministic: re-derivation yields the same PQC public key
    const b = await deriveUnifiedAccount(TEST_MNEMONIC, 0);
    expect(toHex(b.pqc.publicKey)).toBe(toHex(a.pqc.publicKey));
  });

  it("returns 0x-prefixed private (32B) and compressed public (33B) keys", async () => {
    const a = await deriveUnifiedAccount(TEST_MNEMONIC, 0);
    expect(a.privateKey).toMatch(/^0x[0-9a-f]{64}$/);
    expect(a.publicKey).toMatch(/^0x[0-9a-f]{66}$/);
    expect(a.mnemonic).toBe(TEST_MNEMONIC);
  });

  it("varies by index", async () => {
    const a0 = await deriveUnifiedAccount(TEST_MNEMONIC, 0);
    const a1 = await deriveUnifiedAccount(TEST_MNEMONIC, 1);
    expect(a1.cosmos).not.toBe(a0.cosmos);
    expect(isValidBech32(a1.cosmos, "qor")).toBe(true);
  });

  it("rejects an invalid mnemonic and a bad index", async () => {
    await expect(deriveUnifiedAccount("not a mnemonic", 0)).rejects.toThrow(
      /invalid mnemonic/,
    );
    await expect(deriveUnifiedAccount(TEST_MNEMONIC, -1)).rejects.toThrow();
  });
});

describe("unifiedAccountFromSeed", () => {
  it("matches the 0x01×32 known-answer addresses exactly", () => {
    const s = unifiedAccountFromSeed(new Uint8Array(32).fill(1));
    expect(s.cosmos).toBe(KAT_SEED.cosmos);
    expect(s.evm).toBe(KAT_SEED.evm);
    expect(s.svm).toBe(KAT_SEED.svm);
    expect(s.mnemonic).toBeUndefined();
    expect(s.pqc.publicKey.length).toBe(2592);
    // Seed-path PQC determinism (uses the "seed:" prefix in the PQC seed context).
    expect(toHex(s.pqc.publicKey.slice(0, 8))).toBe("2d7f888fecbe5b24");
  });

  it("rejects a non-32-byte seed", () => {
    expect(() => unifiedAccountFromSeed(new Uint8Array(31))).toThrow(/32 bytes/);
  });
});

describe("addressesFrom20 / qoreAddresses", () => {
  it("renders all three encodings from 20 bytes", async () => {
    const a = await deriveUnifiedAccount(TEST_MNEMONIC, 0);
    const enc = addressesFrom20(a.addressBytes);
    expect(enc.cosmos).toBe(KAT0.cosmos);
    expect(enc.evm).toBe(KAT0.evm);
    expect(enc.svm).toBe(KAT0.svm);
  });

  it("rejects a non-20-byte address", () => {
    expect(() => addressesFrom20(new Uint8Array(19))).toThrow(/20 bytes/);
  });

  it("round-trips via any single encoding", () => {
    const fromEvm = qoreAddresses({ evm: KAT0.evm });
    expect(fromEvm.cosmos).toBe(KAT0.cosmos);
    expect(fromEvm.svm).toBe(KAT0.svm);

    const fromCosmos = qoreAddresses({ cosmos: KAT0.cosmos });
    expect(fromCosmos.evm).toBe(KAT0.evm);

    const fromHex = qoreAddresses({ hex: KAT0.addr20 });
    expect(fromHex.cosmos).toBe(KAT0.cosmos);
    expect(fromHex.evm).toBe(KAT0.evm);
  });

  it("throws when no input is given", () => {
    expect(() => qoreAddresses({})).toThrow(/one of/);
  });
});

File: packages/go/qorechain/unified/unified_test.go
package unified

import (
	"bytes"
	"encoding/hex"
	"strings"
	"testing"

	sdktx "github.com/cosmos/cosmos-sdk/types/tx"
	"github.com/cosmos/gogoproto/proto"
	dcrsecp "github.com/decred/dcrd/dcrec/secp256k1/v4"
	dcrecdsa "github.com/decred/dcrd/dcrec/secp256k1/v4/ecdsa"

	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"

	"github.com/qorechain/qorechain-sdk/packages/go/qorechain/signbytes"
)

const katMnemonic = "test test test test test test test test test test test junk"

func TestDeriveUnifiedAccountKAT(t *testing.T) {
	acc, err := DeriveUnifiedAccount(katMnemonic, 0)
	if err != nil {
		t.Fatalf("derive: %v", err)
	}
	if got, want := acc.Cosmos, "qor17w0adeg64ky0daxwd2ugyuneellmjgnxhkv37z"; got != want {
		t.Errorf("cosmos = %s, want %s", got, want)
	}
	if got, want := acc.Evm, "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266"; got != want {
		t.Errorf("evm = %s, want %s", got, want)
	}
	if got, want := acc.Svm, "HQ1S8pxTw4YN41GPdWYPQVfweQveXAUtfFnZNmvPfrYf"; got != want {
		t.Errorf("svm = %s, want %s", got, want)
	}
	if len(acc.Pqc.PublicKey) != 2592 || len(acc.Pqc.SecretKey) != 4896 {
		t.Errorf("pqc lengths = %d/%d, want 2592/4896", len(acc.Pqc.PublicKey), len(acc.Pqc.SecretKey))
	}
	// PQC-pubkey determinism: first 8 bytes lock exact cross-language derivation.
	if got, want := hex.EncodeToString(acc.Pqc.PublicKey[:8]), "4a685622f2a99d54"; got != want {
		t.Errorf("pqc pubkey prefix = %s, want %s", got, want)
	}
	if len(acc.PublicKey) != 33 {
		t.Errorf("compressed pubkey = %d bytes, want 33", len(acc.PublicKey))
	}
	if len(acc.PrivateKey) != 32 {
		t.Errorf("private key = %d bytes, want 32", len(acc.PrivateKey))
	}
}

func TestUnifiedAccountFromSeedKAT(t *testing.T) {
	seed := bytes.Repeat([]byte{0x01}, 32)
	acc, err := UnifiedAccountFromSeed(seed)
	if err != nil {
		t.Fatalf("from seed: %v", err)
	}
	if got, want := acc.Cosmos, "qor1rfjz7r3u8t65teavh5utquj3kwvsj983hh5zaj"; got != want {
		t.Errorf("cosmos = %s, want %s", got, want)
	}
	if got, want := acc.Evm, "0x1a642f0E3c3aF545E7AcBD38b07251B3990914F1"; got != want {
		t.Errorf("evm = %s, want %s", got, want)
	}
	if got, want := acc.Svm, "2n2Cnc7fib6rm4Azo1mbvMuHB3iCb1J254kEQDcrHyPu"; got != want {
		t.Errorf("svm = %s, want %s", got, want)
	}
	if len(acc.Pqc.PublicKey) != 2592 || len(acc.Pqc.SecretKey) != 4896 {
		t.Errorf("pqc lengths = %d/%d, want 2592/4896", len(acc.Pqc.PublicKey), len(acc.Pqc.SecretKey))
	}
	// PQC-pubkey determinism for the seed path (uses the "seed:"+hex(priv) suffix).
	if got, want := hex.EncodeToString(acc.Pqc.PublicKey[:8]), "2d7f888fecbe5b24"; got != want {
		t.Errorf("pqc pubkey prefix = %s, want %s", got, want)
	}
}

func TestDeriveUnifiedDeterministic(t *testing.T) {
	a, err := DeriveUnifiedAccount(katMnemonic, 0)
	if err != nil {
		t.Fatal(err)
	}
	b, err := DeriveUnifiedAccount(katMnemonic, 0)
	if err != nil {
		t.Fatal(err)
	}
	if a.Cosmos != b.Cosmos || !bytes.Equal(a.Pqc.SecretKey, b.Pqc.SecretKey) {
		t.Fatal("derivation is not deterministic")
	}
	c, err := DeriveUnifiedAccount(katMnemonic, 1)
	if err != nil {
		t.Fatal(err)
	}
	if a.Cosmos == c.Cosmos {
		t.Fatal("different indices must yield different accounts")
	}
}

func TestAddressesFrom20AndQoreAddresses(t *testing.T) {
	acc, err := DeriveUnifiedAccount(katMnemonic, 0)
	if err != nil {
		t.Fatal(err)
	}
	// AddressesFrom20 round-trips to the same three encodings.
	enc, err := AddressesFrom20(acc.AddressBytes)
	if err != nil {
		t.Fatal(err)
	}
	if enc.Cosmos != acc.Cosmos || enc.Evm != acc.Evm || enc.Svm != acc.Svm {
		t.Fatal("AddressesFrom20 mismatch")
	}
	// QoreAddresses from each of the three input forms must agree.
	fromEvm, err := QoreAddresses("", acc.Evm, "")
	if err != nil {
		t.Fatal(err)
	}
	fromCosmos, err := QoreAddresses(acc.Cosmos, "", "")
	if err != nil {
		t.Fatal(err)
	}
	fromHex, err := QoreAddresses("", "", hex.EncodeToString(acc.AddressBytes[:]))
	if err != nil {
		t.Fatal(err)
	}
	for _, g := range []Addresses{fromEvm, fromCosmos, fromHex} {
		if g.Cosmos != acc.Cosmos || g.Evm != acc.Evm || g.Svm != acc.Svm {
			t.Fatalf("QoreAddresses mismatch: %+v", g)
		}
	}
}

// TestUnifiedAccountFromPhantomSignatureRemoved asserts the signature-derived
// account constructor is gone: it returns an error and no account, and the error
// points callers at the authenticator lanes.
func TestUnifiedAccountFromPhantomSignatureRemoved(t *testing.T) {
	sig := bytes.Repeat([]byte{0xAB}, 64)
	acc, err := UnifiedAccountFromPhantomSignature(sig)
	if err == nil {
		t.Fatal("expected an error: deriving a spend key from a wallet signature must not be possible")
	}
	if acc.PrivateKey != nil || acc.Cosmos != "" || acc.Evm != "" || acc.Svm != "" {
		t.Fatalf("expected the zero account, got %+v", acc)
	}
	for _, want := range []string{
		"removed in v0.8.0",
		"MsgRegisterAuthenticator",
		"MsgExecuteCosmos",
		"MsgExecuteEVM",
		"treated as exposed",
	} {
		if !strings.Contains(err.Error(), want) {
			t.Fatalf("error message must mention %q, got: %s", want, err.Error())
		}
	}
}

// verifyEthSig checks a 64-byte r‖s signature as secp256k1 over keccak256(msg)
// against the account's compressed public key.
func verifyEthSig(t *testing.T, compressedPub, msg, sig []byte) bool {
	t.Helper()
	if len(sig) != 64 {
		t.Fatalf("signature must be 64 bytes, got %d", len(sig))
	}
	pub, err := dcrsecp.ParsePubKey(compressedPub)
	if err != nil {
		t.Fatalf("parse pubkey: %v", err)
	}
	var r, s dcrsecp.ModNScalar
	if r.SetByteSlice(sig[:32]) {
		t.Fatal("r overflow")
	}
	if s.SetByteSlice(sig[32:]) {
		t.Fatal("s overflow")
	}
	ecdsaSig := dcrecdsa.NewSignature(&r, &s)
	return ecdsaSig.Verify(keccak(msg), pub)
}

func TestSignClassicalEthVerifiesOverKeccak(t *testing.T) {
	acc, err := DeriveUnifiedAccount(katMnemonic, 0)
	if err != nil {
		t.Fatal(err)
	}
	msg := &banktypes.MsgSend{
		FromAddress: acc.Cosmos,
		ToAddress:   acc.Cosmos,
	}
	raw, err := SignClassicalEth(EthSignParams{
		Account:       acc,
		ChainID:       "qorechain-vladi",
		AccountNumber: 7,
		Sequence:      3,
		Messages:      []Message{{TypeURL: "/cosmos.bank.v1beta1.MsgSend", Value: msg}},
		Fee:           Fee{Amount: []Coin{{Denom: "uqor", Amount: "100"}}, Gas: 200000},
	})
	if err != nil {
		t.Fatalf("sign classical: %v", err)
	}
	var txRaw sdktx.TxRaw
	if err := proto.Unmarshal(raw, &txRaw); err != nil {
		t.Fatalf("decode TxRaw: %v", err)
	}
	if len(txRaw.Signatures) != 1 {
		t.Fatalf("expected 1 signature, got %d", len(txRaw.Signatures))
	}
	// Reconstruct the SignDoc the signature must cover.
	signDoc, err := proto.Marshal(&sdktx.SignDoc{
		BodyBytes:     txRaw.BodyBytes,
		AuthInfoBytes: txRaw.AuthInfoBytes,
		ChainId:       "qorechain-vladi",
		AccountNumber: 7,
	})
	if err != nil {
		t.Fatal(err)
	}
	if !verifyEthSig(t, acc.PublicKey, signDoc, txRaw.Signatures[0]) {
		t.Fatal("classical eth signature does not verify as secp256k1 over keccak256(SignDoc)")
	}
	// Classical body must NOT carry a PQC extension.
	var body sdktx.TxBody
	if err := proto.Unmarshal(txRaw.BodyBytes, &body); err != nil {
		t.Fatal(err)
	}
	if len(body.ExtensionOptions) != 0 {
		t.Fatal("classical body must have no extension options")
	}
	// The signer pubkey Any must use the eth_secp256k1 type URL.
	var ai sdktx.AuthInfo
	if err := proto.Unmarshal(txRaw.AuthInfoBytes, &ai); err != nil {
		t.Fatal(err)
	}
	if got := ai.SignerInfos[0].PublicKey.TypeUrl; got != ETHSECP256K1PubKeyType {
		t.Fatalf("pubkey type = %s, want %s", got, ETHSECP256K1PubKeyType)
	}
}

func TestSignHybridEthVerifiesAndFrames(t *testing.T) {
	acc, err := DeriveUnifiedAccount(katMnemonic, 0)
	if err != nil {
		t.Fatal(err)
	}
	msg := &banktypes.MsgSend{FromAddress: acc.Cosmos, ToAddress: acc.Cosmos}
	raw, err := SignHybridEth(EthSignParams{
		SignBytesVersion: signbytes.V1,
		Account:          acc,
		ChainID:          "qorechain-vladi",
		AccountNumber:    7,
		Sequence:         3,
		Messages:         []Message{{TypeURL: "/cosmos.bank.v1beta1.MsgSend", Value: msg}},
		Fee:              Fee{Amount: []Coin{{Denom: "uqor", Amount: "100"}}, Gas: 200000},
	})
	if err != nil {
		t.Fatalf("sign hybrid: %v", err)
	}
	var txRaw sdktx.TxRaw
	if err := proto.Unmarshal(raw, &txRaw); err != nil {
		t.Fatal(err)
	}
	// Final body carries exactly one PQC extension option.
	var body sdktx.TxBody
	if err := proto.Unmarshal(txRaw.BodyBytes, &body); err != nil {
		t.Fatal(err)
	}
	if len(body.ExtensionOptions) != 1 {
		t.Fatalf("hybrid body must have 1 extension option, got %d", len(body.ExtensionOptions))
	}

	// Rebuild B0 (body WITHOUT ext) and verify the ML-DSA-87 signature over
	// frame(B0, authInfo).
	b0, err := proto.Marshal(&sdktx.TxBody{
		Messages:      body.Messages,
		Memo:          body.Memo,
		TimeoutHeight: body.TimeoutHeight,
	})
	if err != nil {
		t.Fatal(err)
	}
	if bytes.Equal(b0, txRaw.BodyBytes) {
		t.Fatal("B0 must differ from the final body (ext must be present in final body only)")
	}

	// Classical signature verifies over keccak256(SignDoc(finalBody, A)).
	signDoc, err := proto.Marshal(&sdktx.SignDoc{
		BodyBytes:     txRaw.BodyBytes,
		AuthInfoBytes: txRaw.AuthInfoBytes,
		ChainId:       "qorechain-vladi",
		AccountNumber: 7,
	})
	if err != nil {
		t.Fatal(err)
	}
	if !verifyEthSig(t, acc.PublicKey, signDoc, txRaw.Signatures[0]) {
		t.Fatal("hybrid classical signature does not verify over keccak256(SignDoc)")
	}
}

func TestParseOnChainPubKey(t *testing.T) {
	acc, err := DeriveUnifiedAccount(katMnemonic, 0)
	if err != nil {
		t.Fatal(err)
	}
	raw, err := SignClassicalEth(EthSignParams{
		Account:       acc,
		ChainID:       "qorechain-vladi",
		AccountNumber: 1,
		Sequence:      0,
		Messages:      []Message{{TypeURL: "/cosmos.bank.v1beta1.MsgSend", Value: &banktypes.MsgSend{FromAddress: acc.Cosmos, ToAddress: acc.Cosmos}}},
		Fee:           Fee{Amount: []Coin{{Denom: "uqor", Amount: "1"}}, Gas: 100000},
	})
	if err != nil {
		t.Fatal(err)
	}
	var txRaw sdktx.TxRaw
	if err := proto.Unmarshal(raw, &txRaw); err != nil {
		t.Fatal(err)
	}
	var ai sdktx.AuthInfo
	if err := proto.Unmarshal(txRaw.AuthInfoBytes, &ai); err != nil {
		t.Fatal(err)
	}
	parsed, err := ParseOnChainPubKeyAny(ai.SignerInfos[0].PublicKey)
	if err != nil {
		t.Fatalf("parse pubkey: %v", err)
	}
	if parsed.Addresses.Cosmos != acc.Cosmos {
		t.Fatalf("parsed cosmos = %s, want %s", parsed.Addresses.Cosmos, acc.Cosmos)
	}
	if !bytes.Equal(parsed.CompressedPubKey, acc.PublicKey) {
		t.Fatal("parsed pubkey mismatch")
	}
}
